Reduce CppInterOp Emscripten shared library size

mcbarton opened the pull request:
# Description

Please include a summary of changes, motivation and context for this PR.

@vgvassilev This change locally reduced the Emscripten shared library size to 47/48 Mb (depends of whether you use dir -lh or du -lh). It still passed all tests and xeus-cpp still could run all the notebook cells it could before (exception throwing is broken in CppInterOps deployment at the moment for some reason). I have cleared the cache of all llvm 20 Emscripten builds for this PR.

Fixes # (issue)

## Type of change

Please tick all options which are relevant.

- [ ] Bug fix
- [ ] New feature
- [x] Requires documentation updates

## Testing

Please describe the test(s) that you added and ran to verify your changes.

## Checklist

- [x] I have read the contribution guide recently

anutosh491 commented:
It's a compile time optimization flag I think. 

Do we need to force it like this everywhere or can we have cmake handle this ?

anutosh491 commented:
Yeah but we still don't want users to look into/pass flags correct during the build correct?

 We should be able to have cmake handle this.

mcbarton commented:
> Yeah but we still don't want users to look into/pass flags correct during the build correct?
> 
> We should be able to have cmake handle this.

I really don't understand the point your trying to make. Using EMCC_CFLAGS in this way to perfectly legitimate way of doing the Emscripten build. Emscripten is designed to be effected by modifying this flag (see https://emscripten.org/docs/tools_reference/emcc.html#environment-variables). If a user changes what we have in the documentation for this flag, and messes something up, then that is on them.

anutosh491 commented:
My point is simple. We already have compile time and link time flags being passed through cmake, I don't see a need to pass this externally.

vgvassilev commented:
I think what @anutosh491 means is that we need to make this flag as part of our default compiler flags in the CMakeLists.txt.

mcbarton commented:
@vgvassilev @anutosh491 I first tried locally using `-DCMAKE_CXX_FLAGS="-Dwait4=__syscall_wait4 -Oz" -DCMAKE_C_FLAGS="-Oz" -DCMAKE_STATIC_LINKER_FLAGS="-Oz"`  (since I thought this might be the closest thing to EMCC_CFLAGS) and it crashed 15% into the build. I then tried `-DCMAKE_CXX_FLAGS="-Dwait4=__syscall_wait4 -Oz" -DCMAKE_C_FLAGS="-Oz"` which although I could build llvm with this option, had no effect on the size of CppInterOps shared library, or the llvm libraries. I do not know how to proceed now.

mcbarton commented:
Putting a note here in case I forget. Using EMCC_CFLAGS="-flto" in the llvm build doesn't end up effecting the shared library size on its own. Used in conjuncation with Oz then the size falls slightly to 45/46 Mb.

mcbarton commented:
I have updated this PR to use the flto flag as well, and to compile+link CppInterOp shared library with these flags too. This reduces the shared library size to 44/45 Mb.

mcbarton commented on .github/workflows/emscripten.yml:
```diff
@@ -194,6 +194,9 @@ jobs:
                         -DLLVM_NATIVE_TOOL_DIR=$NATIVE_DIR 		\
+                        -DCMAKE_C_FLAGS_RELEASE="-Oz -g0" \
+                        -DCMAKE_CXX_FLAGS_RELEASE="-Oz -g0" \
+                        -DLLVM_ENABLE_LTO=Full \
```
We absolutely need to change the llvm build to get this reduction in size of Emscripten shared library. The majority of the size reduction this PR provides comes from adding these flags to the llvm build.

vgvassilev commented:
Can we add some check in the CI to fire when we break this. Maybe something that checks if the file is greater than 50 and fail the build?

mcbarton commented:
> Can we add some check in the CI to fire when we break this. Maybe something that checks if the file is greater than 50 and fail the build?

I have added a check, which on my local system will error if the size is greater than 46 Mb. Need to wait for ci to pass, to check I implemented correctly in the workflow file.

vgvassilev commented:
> Hi, please refrain from merging this PR before a set of reviews on my end.
> 
> Emscripten-forge applies few optimization flags itself (similar to what pyodide does) to add a basic level of optimization for every package while building.
> 
> https://github.com/emscripten-forge/recipes/blob/e04b1067f9d7e4e56d742e5d3efa2131ed8a0315/recipes/recipes/emscripten_emscripten-wasm32/activate.sh#L50-L68
> 
> So if this goes in (and after a release if we build the latest cppinterop on emscripten-forge) It's possible these flags go against each other/ overwrite each other.

I do not see how these flags can go against each other. We have tests that are running and if the tests are good I do not see a reason to hold things off. Right now we destroy people's network by forcing them to download 300mb, ~100 of which is CppInterOp. If this PR cuts the CppInterOp size by half while the tests still run this is something we need urgently, right?

anutosh491 commented:
Hey @mcbarton can you educate me with some questions

1) What's the current size of libclangCppInterOp.so being built on main ? Guessing its somewhere around 70-80 MB

2) Now as we see that CppInterop's deployment runs on iphone .... which means that the above size would suffice somehow, I am trying to understand what exactly is being optimized here ?

Cause according to my attempts 
i) I see that using o2 gets it to 80 MB
ii) Using oz gets it to 73 MB
iii) flto doesn't have any strong affect
iv) even getting rid of the huge list of symbols we export isn't really having any strong affect on size.

So only making changes to cppinterop might not be good enough ... which gets us to llvm. 

Now I am trying to understand what exactly is being optimized through the above flags to get job done cause libclangInterpreter.a which we link against is anyways not too huge (412K on emscripten-forge) though the transitive dependencies being pulled in might be heavy (I see libclang-cpp.a is 147M)

mcbarton commented:
@vgvassilev @anutosh491 If I build locally according to the instructions in the readme of this PR, this it the deployment I get https://mcbarton.github.io/xeus-cpp-demo/lab/index.html.  Everything which works for me when follow mains Emscripten instructions works in this deployment too. 

The 'SIMD Acceleration through WebAssembly' example doesn't work in the link, but it doesn't work locally if you build according to the instructions on main (e.g. no modified llvm or CppInterOp). That is the reason I have opened this issue https://github.com/compiler-research/CppInterOp/issues/680

mcbarton commented:
> @vgvassilev @anutosh491 If I build locally according to the instructions in the readme of this PR, this is the deployment I get https://mcbarton.github.io/xeus-cpp-demo/lab/index.html. Everything which works for me when I follow mains Emscripten instructions works in this deployment too.
> 
> The 'SIMD Acceleration through WebAssembly' example doesn't work in the link (fails to find wasm_simd128.h), but it doesn't work locally if you build according to the instructions on main (e.g. no modified llvm or CppInterOp). That is the reason I have opened this issue #680

With the changes in this PR, the shared library size is down to 39 Mb. The ci has been changed to check whether any changes make the library greater than 40 Mb.

mcbarton commented on .github/workflows/emscripten.yml:
```diff
@@ -175,7 +175,6 @@ jobs:
                         -DLLVM_EXTERNAL_CLING_SOURCE_DIR=../../cling  \
                         -DCMAKE_BUILD_TYPE=Release \
                         -DLLVM_HOST_TRIPLE=wasm32-unknown-emscripten \
-                        -DLLVM_ENABLE_ASSERTIONS=ON                        \
```
I removed LLVM_ENABLE_ASSERTIONS from the build instructions since off is the default for a release build.

vgvassilev replied:
Can we do that in a separate PR?

mcbarton replied:
Out of curiosity why do you want this in a separate PR? This PR is about reducing the shared library size, and assertions are one thing which is causing the shared library to be large. They are responsible for 5 Mb of the size reduction.

vgvassilev commented on lib/CppInterOp/CMakeLists.txt:
```diff
@@ -127,10 +127,16 @@ if(EMSCRIPTEN)
     PROPERTIES NO_SONAME 1
   )
+  target_compile_options(clangCppInterOp
+    PRIVATE "SHELL: -Oz"
+    PRIVATE "SHELL: -flto"
```
Can't these become part of `CMAKE_X_FLAGS_RELEASE` flags, too?

mcbarton replied:
I tried this out and couldn't get it to work. For some reason if I do it via cmake options then Ubuntu both x86 and arm cannot pass xeus-cpps Emscripten tests for llvm 20. I have reverted back to the original solution now, and waiting to see the ci pass

vgvassilev replied:
How's that even possible...

mcbarton commented:
The old way of doing it is still able to pass the ci for all jobs.

anutosh491 commented:
Hi, 

Some remarks based on my comment here (https://github.com/compiler-research/CppInterOp/pull/655#issuecomment-3077043264)

One of my questions there was 

> I am trying to understand what exactly is being optimized here ?

To be fair, I still don't understand what exactly brings the size :\ 

Just thinking about what can bring down the size from 80 to 45

1) Just making changes to cppinterop bring a change of about +- 5MB (which says we need to make changes to llvm too)
2) Changing from 02 to 0z for any project can also probably get a change of +- 5 to 10 MB (these flags are not very vastly different)
3) I can see that LLVM_ENABLE_ASSERTIONS would also bring the size down.
4) which means LTO on llvm does help but I am not sure I can pin point how it does so

I would have been really happy if we could pin point on what exactly changes in llvm when we us lto (in combination with oz) and how does it get down cppinterop's size as a whole but I guess this is OK.

Let's try this approach out and try to backtrack how it helps :)
